### PR TITLE
Sorting of Privileges

### PR DESCRIPTION
**Pull request by pauldurbin**
### What?

Privileges need to be sorted by relevance.  This has been done by generating a score for each VRN and sorting based on that.

* Updating the sorting of privileges to be based on scope scoring.

### CODE REVIEW

File: lib/volcanic/authenticator/v1/scope.rb [hunk 1]
```diff
@@ -44,10 +47,36 @@ def to_s
...
+      # Dataset multiply 2
+      # Resource multiply 3
+      # ResourceId multiply 4
+      # Qualifiers multiply 5
+      def score
+        %i[stack_id dataset_id resource resource_id qualifiers].map do |element|
+          next if send(element).blank?
+
+          if send(element) == '*'
+            0 * multiplier(element)
```
Comment (paulccarey) [inline lib/volcanic/authenticator/v1/scope.rb]: `0 x anything = 0` right? 🤔

File: lib/volcanic/authenticator/v1/scope.rb [hunk 2]
```diff
@@ -44,10 +47,36 @@ def to_s
...
+          elsif send(element) > '*'
```
Comment (paulccarey) [inline lib/volcanic/authenticator/v1/scope.rb]: this is interesting, does the splat operator take on special meaning with the `>` operator?
did a bit of playing around and it basically looks like everything evaluates true but just wondering why you simply didn't do `thing != '*'`

File: lib/volcanic/authenticator/v1/scope.rb (hunk 1 above)
Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/scope.rb]: good point - it's been a long day!

Comment (paulccarey) [reply lib/volcanic/authenticator/v1/scope.rb]: ha ha, exactly the sort of thing I'd do

File: lib/volcanic/authenticator/v1/scope.rb (hunk 2 above)
Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/scope.rb]: Very good point, and on top of that I'm going to refactor out the above issue 👍

Comment (paulccarey) [reply lib/volcanic/authenticator/v1/scope.rb]: cool, was just more interested tbh if there was a special case here

Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/scope.rb]: Regards the splat, it's not actually a splat, just an asterisk as a string 😃

Comment (paulccarey) [reply lib/volcanic/authenticator/v1/scope.rb]: indeed it is... 😉

File: lib/volcanic/authenticator/v1/privilege.rb [hunk 3]
```diff
@@ -22,13 +22,23 @@ def initialize(scope:, permission_id: nil, group_id: nil, allow:, **args)
         @allow = allow
       end
 
+      def <=>(other)
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/privilege.rb]: I *think* this is returning the most specific privilege scope last with true after false.

this means that given the privileges:
```ruby
{ scope: 'vrn:*:*:jobs/*', allow: false },
{ scope: 'vrn:*:*:jobs/*?creator=me', allow: true },
{ scope: 'vrn:*:*:jobs/23', allow: true },
{ scope: 'vrn:*:*:jobs/*?created_before=tuesday', allow: false }
```

would return:
```ruby
{ scope: 'vrn:*:*:jobs/*', allow: false },
{ scope: 'vrn:*:*:jobs/23', allow: false },
{ scope: 'vrn:*:*:jobs/24', allow: true }
```

This *probably* isn't an issue, but it is worth noting that where the scopes are more specific the most important is the last one, but when they are equally specific the most important (deny) may not be last

File: lib/volcanic/authenticator/v1/scope.rb [hunk 4]
```diff
@@ -44,10 +47,38 @@ def to_s
...
+      def score
+        %i[stack_id dataset_id resource resource_id qualifiers].map do |element|
+          next if send(element).blank?
+
+          if send(element) != '*'
+            0.5 * multiplier(element)
+          else
+            0
+          end
+        end.reject(&:blank?).reduce(&:+)
```
[the remark this answers is not in the document]
Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/scope.rb]: Nope, but updated and will be in the next push.

File: lib/volcanic/authenticator/v1/privilege.rb (hunk 3 above)
Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/privilege.rb]: Would it be a good idea to list some more test cases - I'll add them to the project and we can be sure?

Comment (danleyden) [reply lib/volcanic/authenticator/v1/privilege.rb]: https://volcanicuk.atlassian.net/wiki/spaces/ENG/pages/413892609/Auth+service has a lot of worked examples which include details of which privileges would be returned by the auth service and which should be ultimately applied and reasons why... Are more examples than those provided there needed?

Comment (pauldurbin) [reply lib/volcanic/authenticator/v1/privilege.rb]: I'll pick off a few and add them in to be sure.

File: lib/volcanic/authenticator/v1/scope.rb [hunk 5]
```diff
@@ -44,10 +47,37 @@ def to_s
...
+      # element = * (0)
+      # element is value (0.5)
+      # Stack multiply 1
+      # Dataset multiply 2
+      # Resource multiply 3
+      # ResourceId multiply 4
+      # Qualifiers multiply 5
+      def score
+        %i[stack_id dataset_id resource resource_id qualifiers].reduce(0) do |acc, element|
+          val = if !send(element).blank? && send(element) > '*'
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/scope.rb]: `>` looks an odd operator - is that really correct? should it be != ?

File: lib/volcanic/authenticator/v1/scope.rb [hunk 6]
```diff
@@ -44,10 +47,37 @@ def to_s
...
+
+      # For each element in the VRN, assign it a value
+      # element = * (0)
+      # element is value (0.5)
+      # Stack multiply 1
+      # Dataset multiply 2
+      # Resource multiply 3
+      # ResourceId multiply 4
+      # Qualifiers multiply 5
+      def score
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/scope.rb]: can this be renamed? I fear that a method named "score" is likely to be confused....

consider `specificity_score`

File: lib/volcanic/authenticator/v1/scope.rb [hunk 7]
```diff
@@ -44,10 +47,37 @@ def to_s
         base
       end
 
+      def <=>(other)
+        score < other.score
+      end
+
+      # For each element in the VRN, assign it a value
+      # element = * (0)
+      # element is value (0.5)
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/scope.rb]: 0.5 no longer appears

...and the rest aren't multipliers.... may want renaming

File: lib/volcanic/authenticator/v1/privilege.rb [hunk 8]
```diff
@@ -22,13 +22,27 @@ def initialize(scope:, permission_id: nil, group_id: nil, allow:, **args)
...
+        if scope == other.scope
+          allow_bool <=> other.allow_bool
+        else
+          Scope.parse(scope).specificity_score <=> Scope.parse(other.scope).specificity_score
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/privilege.rb]: why is `scope` not an instance of `Scope`? surely that would be more coherent?

also.... `scope <=> other.scope` gives the same value as `scope.specificity_score <=> other.scope.specificity_score` but with better encapsulation - let `Scope` figure out how to compare specificity with another `Scope`

File: lib/volcanic/authenticator/v1/privilege.rb [hunk 9]
```diff
@@ -22,13 +22,27 @@ def initialize(scope:, permission_id: nil, group_id: nil, allow:, **args)
         @allow = allow
       end
 
+      def <=>(other)
+        if scope == other.scope
```
Comment (danleyden) [inline lib/volcanic/authenticator/v1/privilege.rb]: this logic is still not quite correct.

```ruby
privilege = {scope: 'vrn:*:*:jobs/23', allow: false }
other = { scope: 'vrn:*:*:jobs/224527', allow: true }

scope == other.scope # false
scope <=> other.scope # = 0

[privilege, other].sort == [other, privilege].sort # false - non-deterministic sorting.
```

**Comment by danleyden** [comment]
needs squashing please

**Comment by pauldurbin** [comment]
I always squash and rebase once approved.  Cheers for your feedback on this one 👍

**Pull request merged**